Introduce `mul_by_inverse_unchecked`, and use it

Pratyush opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before hitting that submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

Resolves https://github.com/arkworks-rs/r1cs-std/pull/70#issuecomment-874472795.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

Pratyush commented on src/fields/mod.rs:
```diff
@@ -158,8 +158,33 @@ pub trait FieldVar<F: Field, ConstraintF: Field>:
     fn mul_by_inverse(&self, d: &Self) -> Result<Self, SynthesisError> {
-        let d_inv = d.inverse()?;
-        Ok(d_inv * self)
+        // Enforce that `d` is not zero.
+        d.enforce_not_equal(&Self::zero())?;
+        self.mul_by_inverse_unchecked(d)
```
@weikengchen I've changed this to first enforce that `d != zero`, and then invoking `mul_by_inverse_unchecked`.

weikengchen replied:
Agree!

Pratyush commented on src/fields/mod.rs:
```diff
@@ -158,8 +158,33 @@ pub trait FieldVar<F: Field, ConstraintF: Field>:
+    /// This method *does not* check if `d == 0`.
+    fn mul_by_inverse_unchecked(&self, d: &Self) -> Result<Self, SynthesisError> {
+        let cs = self.cs().or(d.cs());
+        match cs {
+            // If we're in the constant case, we just allocate a new constant having value equalling
+            // `self * d.inverse()`.
+            ConstraintSystemRef::None => Self::new_constant(
+                cs,
+                self.value()? * d.value()?.inverse().expect("division by zero"),
+            ),
+            // If not, we allocate `result` as a new witness having value `self * d.inverse()`,
+            // and check that `result * d = self`.
+            _ => {
+                let result = Self::new_witness(ark_relations::ns!(cs, "self  * d_inv"), || {
+                    Ok(self.value()? * &d.value()?.inverse().unwrap_or(F::zero())?)
+                })?;
+                result.mul_equals(d, self)?;
+                Ok(result)
+            }
+        }
```
This method allocates `result` with value `self * d.inverse()`, and then checks that `result * d == self`.

Pratyush commented on src/poly/evaluations/univariate/mod.rs:
```diff
@@ -175,10 +175,11 @@ impl<F: PrimeField> EvaluationsVar<F> {
 
-        let lhs = lhs_numerator.mul_by_inverse(&lhs_denominator)?;
+        let lhs = lhs_numerator.mul_by_inverse_unchecked(&lhs_denominator)?;
 
         // `rhs` for coset element `a` is \frac{1}{alpha * a^{-1} - 1} = \frac{1}{alpha * offset^{-1} * a'^{-1} - 1}
-        let alpha_coset_offset_inv = interpolation_point.mul_by_inverse(&self.domain.offset)?;
+        let alpha_coset_offset_inv =
+            interpolation_point.mul_by_inverse_unchecked(&self.domain.offset)?;
 
```
@ValarDragon @tsunrise is it okay to use `unchecked` here?

Pratyush commented on src/poly/evaluations/univariate/mod.rs:
```diff
@@ -190,7 +191,7 @@ impl<F: PrimeField> EvaluationsVar<F> {
             // alpha * offset^{-1} * a'^{-1} - 1
             let lag_donom = &alpha_coset_offset_inv * subgroup_point_inv - F::one();
-            let lag_coeff = lhs.mul_by_inverse(&lag_donom)?;
+            let lag_coeff = lhs.mul_by_inverse_unchecked(&lag_donom)?;
```
Ditto

tsunrise replied:
Can we instead use `mul_by_inverse` here? I did not find any clue that `alpha * offset^{-1} * a'^{-1} != 1`... @ValarDragon is there any guarantee that the highlighted part (in fractal paper p64) is nonzero? 
![image](https://user-images.githubusercontent.com/22125551/124860811-3e03e400-df67-11eb-9908-c13cae611959.png)

Pratyush replied:
ping @ValarDragon

I can revert the domain-specific changes for the time being, if the mul-by-inverse is not performance critical. (I think it isn't, right? As the performance should be dominated by hashes)

ValarDragon replied:
The denominator being zero requires `X = eps * a*` or in other words, the interpolation point to be within a given queries coset. The interpolation point used in FRI is the `alpha` for folding. 

Alpha can't be chosen within `L`, but my recollection is that we don't check for it and instead call it an `|L|/|F|` soundness error (Enforcing that `Pr(alpha \in L) = |L|/|F| < 2^{-128}`). 

So then we only have a soundness concern if we imagined being in a 128 bit or less field, and were not use the right out of domain sampling techniques for FRI. But in this case its not clear what the protocol is, since generic parallel repetition does not work for many round protocol.

Also btw, plz ping me on slack if you'd like my review on things! My github notifications are currently overwhelmed -- having trouble staying on top of this.

ValarDragon replied:
I also think its also fine to not use unchecked here, its a minor efficiency gain. (I'd guess ~200 constraints for the entire circuit)

Pratyush commented:
Do you have a good way to add a regression test for this behaviour?

weikengchen commented on src/fields/mod.rs:
```diff
@@ -158,8 +158,33 @@ pub trait FieldVar<F: Field, ConstraintF: Field>:
+
+    /// Returns `(self / d)`.
+    /// The constraint system will be unsatisfiable when `d = 0`.
+    /// This method *does not* check if `d == 0`.
```
I think we need to change this:
```
/// The constraint system will be unsatisfiable when `d = 0`.
```

Since for this function, it does not exclude the case `d = 0` and `self = 0`, so maybe we can only say that it is unexpected behavior.

Pratyush replied:
Updated the comment. Lemme know if it looks good.

weikengchen replied:
Maybe we should still drop the sentence "/// The constraint system will be unsatisfiable when `d = 0`." as the constraint system would be satisfied indeed when `d = 0 & self = 0`. And I think `/// This method *does not* check if `d == 0`.` suffices.

The PR looks good.

Pratyush replied:
I think the snippet is outdated; see here for the updated comment: https://github.com/arkworks-rs/r1cs-std/blob/4a1823383d571910ac03a2a45ab97a48765ef323/src/fields/mod.rs#L167

weikengchen replied:
Got it. Looks good.

tsunrise reviewed (commented):
looks good to me (though we need to make sure `domain.offset != 0` if we plan to use `unchecked`)

Pratyush commented:
Ok I added a check to make sure `domain.offset` is non-zero by construction. I don't know if this is a good idea: is there ever a case where domain.offset is supposed to be zero? If so, then I can change it to enforce the check inside `lagrange_interpolate_with_non_constant_offset`, instead of in `Radix2DomainVar::new`

ValarDragon commented:
domain.offset should never be zero for multiplicative domains, but I don't think that needs to be proven in circuit. Thats about the circuit definition being valid. (The circuit definition has defined an invalid codeword domain if its zero, the offset derivation shouldn't allow construction of zero domain offsets)

Pratyush commented:
I think @tsunrise's code require `offset` to be a variable, so we have to enforce non-zeroness at some point.

ValarDragon commented:
Oh I see. If necessary as a temporary defense in depth for correctness seems like an alright fix then.

What I'd rather see is we assert this on the DomainVar itself (that its offset and generator are non-zero). Basically in the constructor for DomainVar, include an argument for the inputs being guaranteed to be non-zero. If false, then do the non-zero checks once there.

Pratyush commented:
Yeah, that's what I changed it to do: https://github.com/arkworks-rs/r1cs-std/blob/213e53d6c33fc3eb4c77791a90dc8b5bb5f549aa/src/poly/domain/mod.rs#L28

ValarDragon commented:
I meant explicitly including an argument to not get a constraint created. Currently it seems like a constraint is always created if offset is not a constant?

Pratyush commented:
Ah so like a `new_unchecked` method?

ValarDragon commented:
Its constructed once per query per round, so ~200 times.

Pratyush commented:
I see, with a different offset each time?

ValarDragon commented on src/poly/evaluations/univariate/mod.rs:
```diff
@@ -190,6 +198,7 @@ impl<F: PrimeField> EvaluationsVar<F> {
             debug_assert_eq!(subgroup_points[i] * subgroup_point_inv, F::one());
             // alpha * offset^{-1} * a'^{-1} - 1
             let lag_donom = &alpha_coset_offset_inv * subgroup_point_inv - F::one();
+            // lag_donom might be zero, so should use `mul_by_inverse`.
             let lag_coeff = lhs.mul_by_inverse(&lag_donom)?;
```
Actually, in hindsight the right thing to do here is do `mul_by_inverse_unchecked` still, and if we are worried about this make a constraint for make a constraint for `lhs_numerator` not being `0`. That is equisatisfiable, saves `domain_size - 1` constraints, and is conceptually simpler imo.

ValarDragon replied:
Proof of equisatisfiability.

Denominator is `alpha - {domain point}`. This is `0` iff `alpha \in domain`. `alpha \in domain` iff `Z_{domain}(alpha) = 0`.

`lhs_numerator = alpha^{|domain|} - domain_offset^{|domain|} = Z_{domain}(alpha)`

Pratyush replied:
Good point; fixed as follows:

* Enforce numerator is non-zero.
* Add comment explaining why we can use `unchecked`, and used `unchecked`.

Pratyush commented:
I think I addressed all comments, except for the `new_unchecked` for `DomainVar`. I think for the time being it's fine to use `new`, as it only adds a couple hundred constraints, which would be dominated by the rest of the circuit.

Technically, this is a breaking change due to the changes to the `DomainVar` API, but since the only consumer of that is `ark-ldt`, I believe, I think it's fine to do that.

(Maybe we should have "in-flux" features hidden behind a `unstable` feature flag?)
